p2p/peer.go

```diff
@@ -106,6 +215,126 @@ type PeerUpdate struct {
 	Status PeerStatus
 }
 
+// peerStore manages information about peers. It is currently a bare-bones
+// in-memory store of peer addresses, and will be fleshed out later.
+//
+// The main function of peerStore is currently to dispense peers to connect to
+// (via peerStore.Dispense), giving the caller exclusive "ownership" of that
+// peer until the peer is returned (via peerStore.Return). This is used to
+// schedule and synchronize peer dialing and accepting in the Router, e.g.
+// making sure we only have a single connection (in either direction) to peers.
+type peerStore struct {
+	mtx     sync.Mutex
+	peers   map[NodeID]*storePeer
+	claimed map[NodeID]bool
+}
+
+// newPeerStore creates a new peer store.
+func newPeerStore() *peerStore {
+	return &peerStore{
+		peers:   map[NodeID]*storePeer{},
+		claimed: map[NodeID]bool{},
+	}
+}
+
+// Add adds a peer to the store, given as an address.
+func (s *peerStore) Add(address PeerAddress) error {
+	if err := address.Validate(); err != nil {
+		return err
+	}
+	peerID := address.NodeID()
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	peer, ok := s.peers[peerID]
+	if !ok {
+		peer = newStorePeer(peerID)
+		s.peers[peerID] = peer
+	} else if s.claimed[peerID] {
+		// FIXME: We need to handle modifications of claimed peers somehow.
+		return fmt.Errorf("peer %q is claimed", peerID)
+	}
+	peer.AddAddress(address)
+	return nil
+}
+
+// Claim claims a peer. The caller has exclusive ownership of the peer, and must
+// return it by calling Return(). Returns nil if the peer could not be claimed.
+// If the peer is not known to the store, it is registered and claimed.
+func (s *peerStore) Claim(id NodeID) *storePeer {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	if s.claimed[id] {
+		return nil
+	}
+	peer, ok := s.peers[id]
+	if !ok {
+		peer = newStorePeer(id)
+		s.peers[id] = peer
+	}
+	s.claimed[id] = true
+	return peer
+}
+
+// Dispense finds an appropriate peer to contact and claims it. The caller has
+// exclusive ownership of the peer, and must return it by calling Return(). The
+// peer will not be dispensed again until returned.
+//
+// Returns nil if no appropriate peers are available.
+func (s *peerStore) Dispense() *storePeer {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	for key, peer := range s.peers {
+		switch {
+		case len(peer.Addresses) == 0:
+		case s.claimed[key]:
+		default:
+			s.claimed[key] = true
+			return peer
+		}
+	}
+	return nil
+}
+
+// Return returns a claimed peer, making it available for other
+// callers to claim.
+func (s *peerStore) Return(id NodeID) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	delete(s.claimed, id)
+}
+
+// storePeer is a peer stored in the peerStore.
+//
+// FIXME: This should be renamed peer or something else once the old peer is
+// removed.
+type storePeer struct {
+	ID        NodeID
+	Addresses []PeerAddress
+}
+
+// newStorePeer creates a new storePeer.
+func newStorePeer(id NodeID) *storePeer {
+	return &storePeer{
+		ID:        id,
+		Addresses: []PeerAddress{},
```

Comment:
Why not make this a map?

Reply:
We may want to keep the order here. This is e.g. given by the configuration file, where users may expect us to try addresses in the order they're given. But I'll reconsider this when I get started on peer state management.